Refactoring to enable performance optimizations

jenatali opened the pull request:
This PR removes the dependency on the D3D12TranslationLayer external library. Instead, the commits here import the library as-is, and then strip out the unneeded pieces. There are a lot of them...

The primary goal of this PR was b492fe6 ("Use root UAVs for descriptor space 1") to pair with this compiler change: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26803/diffs?commit_id=1677b86b895871e1b4d16fa2f0be443fe863e70d. Additional changes on top of that to enable more caching weren't planned, but do help with repeated runs of the same app.

Overall performance wins are small but probably worthwhile.

vdwtanner commented on src/openclon12/device.cpp:
```diff
@@ -599,14 +600,9 @@ void Device::FlushAllDevices(TaskPoolLock const& Lock)
-//}
-
-void D3DDevice::ExecuteTasks(Submission& tasks)
+void D3DDevice::ExecuteTasks(std::unique_ptr<Submission> spTasks)
```
What does the `s` prefix stand for here? At first I thought it was a typo, but then saw that you have it in a couple places so probably intentional. The ones that come to mind don't seem to fit here.

jenatali replied:
Heh, it's a bit of a remnant from when I cared more about Hungarian notation I guess. `p` for pointer, `sp` for smart pointer.

vdwtanner commented on src/openclon12/kernel_tasks.cpp:
```diff
@@ -14,15 +14,25 @@
-auto Program::SpecializationKey::Allocate(D3DDevice const* Device, CompiledDxil::Configuration const& conf) -> std::unique_ptr<SpecializationKey>
+size_t Program::SpecializationKey::AllocatedByteSize(uint32_t NumArgs)
+{
+    uint32_t NumAllocatedArgs = NumArgs ? NumArgs - 1 : 0;
```
I realize that this is just bringing forward what was already in the code, but why do we only allocate n-1 args? Shouldn't we have space allocated for all of them?

jenatali replied:
Because the struct ends with an array of size 1, because 0-sized arrays are a compiler extension / unsupported in MSVC C++ I think.

vdwtanner reviewed (approved):
Generally approve since I know you've already done testing, but mainly curious around `GetDXCoreBudget` comment and possibility of `AllocatedByteSize` being off by 1 (it's probably accurate, but I wonder if the actual allocation is off by 1 then?)

jenatali commented on include/d3d12translationlayer/CommandListManager.hpp:
```diff
@@ -58,13 +57,12 @@ namespace D3D12TranslationLayer
         Fence* GetFence() { return &m_Fence; }
 
-        ID3D12GraphicsCommandList* GetGraphicsCommandList(ID3D12CommandList *pCommandList = nullptr) { return  m_type == COMMAND_LIST_TYPE::GRAPHICS ? static_cast<ID3D12GraphicsCommandList * const>(pCommandList ? pCommandList : m_pCommandList.get()) : nullptr; }
+        ID3D12GraphicsCommandList* GetGraphicsCommandList(ID3D12CommandList *pCommandList = nullptr) { return static_cast<ID3D12GraphicsCommandList * const>(pCommandList ? pCommandList : m_pCommandList.get()); }
```
`m_type` no longer exists, which is the thing that was checked to cause the null return. The only command list manager instance is a "graphics" (actually compute) command list manager now.

jenatali commented on include/d3d12translationlayer/ImmediateContext.hpp:
```diff
@@ -1028,71 +960,25 @@ class ImmediateContext
+        if (ComputeOnly() || Type == D3D12_HEAP_TYPE_DEFAULT)
         {
-            return CD3DX12_HEAP_PROPERTIES(Type, GetNodeMask(), GetNodeMask());
+            return CD3DX12_HEAP_PROPERTIES(Type, 1, 1);
```
CL does have a way of selecting devices, but currently CLOn12 only supports "logical" devices, i.e. an entire `IDXGIAdapter` or `IDXCoreAdapter`. The CL API that would maybe be closest to supporting linked adapters would be https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_partitioning_a_device, but I don't think we have a need to support anything interesting there.

D3D11On12 only supports non-zero nodes (non-1 node masks) because our D3D11-based tests use it to run on non-zero nodes, just to ensure that path is tested. It doesn't support multi-node execution. Keeping that logic for CL doesn't seem useful.

jenatali commented on src/d3d12translationlayer/Residency.cpp:
```diff
@@ -261,10 +260,9 @@ HRESULT ResidencyManager::ProcessPagingWork(UINT CommandListIndex, ResidencySet
-static void GetDXCoreBudget(IDXCoreAdapter *AdapterDXCore, UINT NodeIndex, DXCoreAdapterMemoryBudget *InfoOut, DXCoreSegmentGroup Segment)
+static void GetDXCoreBudget(IDXCoreAdapter *AdapterDXCore, DXCoreAdapterMemoryBudget *InfoOut, DXCoreSegmentGroup Segment)
 {
     DXCoreAdapterMemoryBudgetNodeSegmentGroup InputParams = {};
```
It's implicitly 0'd.

vdwtanner replied:
Ah yep, that makes sense with fresh eyes.

jenatali commented:
I rebased, rearranged, and dropped the root UAVs change, since that's an unproven design change. The rest of this is relatively mechanical or straightforward. I've verified that this still works correctly against the CLOn12 compiler from Mesa `main` without the paired root UAV changes. I'll post the root UAV change as a separate PR.
